File: LeetCode/1354. Construct Target Array With Multiple Sums/solution.py
class Solution:
    # def isPossible(self, target: List[int]) -> bool:
    #     sums = sum(target)
    #     target = [-t for t in target]
    #     heapq.heapify(target)
    #     while target:
    #         # print(target)
    #         cur = -heapq.heappop(target)
    #         if cur > 1:
    #             remains = sums - cur
    #             if cur <= remains or remains <= 0:
    #                 return False
    #             cur = cur%remains or remains
    #             sums = remains + cur
    #             heapq.heappush(target,-cur)
    #     return True
    
    def isPossible(self, A):
        total = sum(A)
        A = [-a for a in A]
        heapq.heapify(A)
        while True:
            a = -heapq.heappop(A)
            total -= a
            if a == 1 or total == 1: return True
            if a < total or total == 0 or a % total == 0:
                return False
            a %= total
            total += a
            heapq.heappush(A, -a)

File: LeetCode/1354. Construct Target Array With Multiple Sums/solution.cpp
using ll = long long;
class Solution {
public:
    bool isPossible(vector<int>& target) {
        priority_queue<int,vector<int>> pq;
        ll sums = 0;
        for (auto x : target){
            sums += x;
            pq.push(x);
        }
        while (1){
            int a = pq.top(); pq.pop();
            int r = sums-a;
            if (r == 1 || a == 1)
                return true;
            // cout << a << " " << sums << endl;
            
            if (r == 0 || sums%r == 0 || sums%r == a)
                return false;
            pq.push(sums%r);
            sums = sums%r + r;
        }
        return true;

    }
};